draft-ietf-httpbis-header-structure.md

```diff
@@ -151,70 +152,376 @@ If a member of barURLs is not a valid URI-reference, it MUST cause
+
+Strings are zero or more printable ASCII {{!RFC0020}} characters (i.e., the range 0x20 to 0x7E). Note that this excludes tabs, newlines, carriage returns, etc.
+
+The ABNF for strings is:
+
+~~~ abnf
+sh-string = DQUOTE *(chr) DQUOTE
+chr       = unescaped / escaped
+unescaped = %x20-21 / %x23-5B / %x5D-7E
+escaped   = "\" ( DQUOTE / "\" )
+~~~
+
+In HTTP/1 headers, strings are delimited with double quotes, using a backslash ("\\") to escape double quotes and backslashes. For example:
+
+~~~ example
+Example-StringHeader: "hello world"
+~~~
+
+Note that strings only use DQUOTE as a delimiter; single quotes do not delimit strings. Furthermore, only DQUOTE and "\\" can be escaped; other sequences MUST cause parsing to fail.
+
+Unicode is not directly supported in this document, because it causes a number of interoperability issues, and -- with few exceptions -- header values do not require it.
+
+When it is necessary for a field value to convey non-ASCII string content, binary content ({{binary}}) SHOULD be specified, along with a character encoding (preferably, UTF-8).
+
+Parsers MUST support strings with at least 1024 characters.
+
+
+## Identifiers {#identifier}
+
+Identifiers are short textual identifiers; their abstract model is identical to their expression in the textual HTTP serialisation. Parsers MUST support identifiers with at least 64 characters.
+
+The ABNF for identifiers is:
+
+~~~ abnf
+identifier = lcalpha *( lcalpha / DIGIT / "_" / "-"/ "*" / "/" )
+lcalpha    = %x61-7A ; a-z
+~~~
+
+Note that identifiers can only contain lowercase letters.
+
+
+## Binary Content {#binary}
+
+Arbitrary binary content can be conveyed in Structured Headers.
+
+The ABNF for binary content is:
+
+~~~ abnf
+sh-binary = "*" *(base64) "*"
+base64    = ALPHA / DIGIT / "+" / "/" / "="
+~~~
+
+In HTTP/1 headers, binary content is delimited with asterisks and encoded using base64 ({{!RFC4648}}, Section 4). For example:
+
+~~~ example
+Example-BinaryHdr: *cHJldGVuZCB0aGlzIGlzIGJpbmFyeSBjb250ZW50Lg==*
+~~~
+
+Parsers MUST support binary content with at least 16384 octets after decoding.
+
+
+
+# Structured Headers in HTTP/1 {#text}
+
+This section defines how to serialise and parse Structured Headers in HTTP/1 textual header fields, and protocols compatible with them (e.g., in HTTP/2 {{?RFC7540}} before HPACK {{?RFC7541}} is applied).
+
+## Serialising Structured Headers into HTTP/1 {#text-serialise}
+
+Given a structured defined in this specification:
+
+1. If the structure is a dictionary, let output be the result of Serialising a Dictionary {#ser-dictionary}.
+2. If the structure is a list, let output be the result of Serialising a List {#ser-list}.
+3. If the structure is a parameterised list, let output be the result of Serialising a Parameterised List {#ser-param-list}.
+4. If the structure is an item, let output be the result of Serialising an Item {#ser-item}.
+5. If the structure is anything else (including an identifier), fail serialisation.
```

Comment:
This works, but it seems cleaner as an Assert upfront, since it's invalid input and only other specifications would call this and they better do the correct thing.

Reply:
ack.